# Titanic dataset

Load the titanic dataset using seaborn. pip install seaborn, 'import seaborn as sns', then load using 'sns.load_dataset('titanic')'


In [2]:
#initialisation
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
titanic_df = sns.load_dataset("titanic")
titanic_df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

## Clean the dataset: 

- Where are the missing values, can they be filled/need to be dropped?
    - df.isnull()
    - df.fillna()
    - df.dropna() 
- Are there any duplicates?
    - df.duplicated()
    - df.drop_duplicates()
- Do any columns need renaming?
    - df.rename() 
- Any other data quality concepts you can check?

In [4]:
#python
titanic_df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

We notice that there are 688 nulls for deck, which is a very significant amount. So we drop the deck column.

In [5]:
titanic_df = titanic_df.drop(columns=["deck"])
titanic_df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64

We can drop the rows where embarked and embark_town is null

In [6]:
titanic_df = titanic_df.dropna(subset=['embarked','embark_town'])
titanic_df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
embark_town      0
alive            0
alone            0
dtype: int64

Finally, we see there is a singificant number of nulls for age. We can either fill these with the median value, or drop these rows. Why not both!

In [7]:
titanic_fill_age_df = titanic_df.copy()
titanic_no_fill_age_df = titanic_df.copy()

titanic_fill_age_df['age'] = titanic_fill_age_df['age'].fillna(value=titanic_df['age'].median())
titanic_no_fill_age_df = titanic_no_fill_age_df.dropna()

In [8]:
print(f"The data frame with filled ages has {titanic_fill_age_df.shape[0]} entries and {titanic_fill_age_df.isna().sum().sum()} null values")
print(f"The data frame with filled ages has {titanic_no_fill_age_df.shape[0]} entries and {titanic_no_fill_age_df.isna().sum().sum()} null values")

The data frame with filled ages has 889 entries and 0 null values
The data frame with filled ages has 712 entries and 0 null values


## Check the descriptive statistics
- df.describe()

In [9]:
titanic_fill_age_df.describe()

survived      pclass         age       sibsp       parch        fare
count  889.000000  889.000000  889.000000  889.000000  889.000000  889.000000
mean     0.382452    2.311586   29.315152    0.524184    0.382452   32.096681
std      0.486260    0.834700   12.984932    1.103705    0.806761   49.697504
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   22.000000    0.000000    0.000000    7.895800
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   35.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## Find the average fare and age for each class 
- df.groupby()
- mean()

In [10]:
titanic_no_fill_age_df[['class','age','fare']].groupby(['class'],observed=True).mean()


age       fare
class                       
First   38.105543  88.048121
Second  29.877630  21.471556
Third   25.140620  13.229435

Here we see the average (mean) age and fare for each class.

## Add a new column showing the fare as a percentage of total

In [11]:
titanic_no_fill_age_df['percentage_fare_of_total'] = titanic_no_fill_age_df.fare/titanic_no_fill_age_df.fare.sum()
titanic_no_fill_age_df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  percentage_fare_of_total  
0      man        True  Southampton    no  False                  0.000295  
1    woman       False    Cherbourg   yes  False                  0.002896  
2    woman       False  Southampton   yes   True                  0.000322  
3    woman       False  Southampton   yes  False                  0.002157  
4      man        True  Southampton    no   True                  0.000327  
..     ...         ...          ...   ...    ...                       ...  
885  woman       False   Queenstown    no  False                  0.001183  
886    man        True  Southampton    no   True                  0.000528  
887  woman       False  Southampton   yes   True                  0.001219  
889    man        True    Cherbourg   yes   True                  0.001219  
890    man        True   Queenstown    no   True                  0.000315  

[712 rows x 15 columns]

## How many survivors were female compared to male?

In [16]:
titanic_no_fill_age_df.loc[(titanic_no_fill_age_df['survived']==1) & (titanic_no_fill_age_df['age'] < 18)].groupby(['sex'])[['survived']].count()

survived
sex             
female        38
male          23

There were 102 more female survivors than male

## What interesting insights can you draw?

Let's see if we can find the gender distribution of total surivors by age category

In [17]:
new_df = titanic_no_fill_age_df[['survived','sex','age']].copy()

In [18]:
new_df

survived     sex   age
0           0    male  22.0
1           1  female  38.0
2           1  female  26.0
3           1  female  35.0
4           0    male  35.0
..        ...     ...   ...
885         0  female  39.0
886         0    male  27.0
887         1  female  19.0
889         1    male  26.0
890         0    male  32.0

[712 rows x 3 columns]

In [20]:
bins = [0,2,4,13,18,30,45,65,110]
labels = ['Infant','Toddler','Child','Teen','Young Adult','Middle-Aged','Older Adult','OAP']
new_df['Age Group'] = pd.cut(new_df['age'], bins=bins, labels=labels, right=False)

In [23]:
new_df.loc[new_df['survived']==1]

survived     sex   age    Age Group
1           1  female  38.0  Middle-Aged
2           1  female  26.0  Young Adult
3           1  female  35.0  Middle-Aged
8           1  female  27.0  Young Adult
9           1  female  14.0         Teen
..        ...     ...   ...          ...
875         1  female  15.0         Teen
879         1  female  56.0  Older Adult
880         1  female  25.0  Young Adult
887         1  female  19.0  Young Adult
889         1    male  26.0  Young Adult

[288 rows x 4 columns]

In [26]:
pd.get_dummies(new_df)

survived   age  sex_female  sex_male  Age Group_Infant  \
0           0  22.0       False      True             False   
1           1  38.0        True     False             False   
2           1  26.0        True     False             False   
3           1  35.0        True     False             False   
4           0  35.0       False      True             False   
..        ...   ...         ...       ...               ...   
885         0  39.0        True     False             False   
886         0  27.0       False      True             False   
887         1  19.0        True     False             False   
889         1  26.0       False      True             False   
890         0  32.0       False      True             False   

     Age Group_Toddler  Age Group_Child  Age Group_Teen  \
0                False            False           False   
1                False            False           False   
2                False            False           False   
3                False            False           False   
4                False            False           False   
..                 ...              ...             ...   
885              False            False           False   
886              False            False           False   
887              False            False           False   
889              False            False           False   
890              False            False           False   

     Age Group_Young Adult  Age Group_Middle-Aged  Age Group_Older Adult  \
0                     True                  False                  False   
1                    False                   True                  False   
2                     True                  False                  False   
3                    False                   True                  False   
4                    False                   True                  False   
..                     ...                    ...                    ...   
885                  False                   True                  False   
886                   True                  False                  False   
887                   True                  False                  False   
889                   True                  False                  False   
890                  False                   True                  False   

     Age Group_OAP  
0            False  
1            False  
2            False  
3            False  
4            False  
..             ...  
885          False  
886          False  
887          False  
889          False  
890          False  

[712 rows x 12 columns]